hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3LocatedFileStatus.java

```diff
@@ -0,0 +1,58 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ * <p>
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hadoop.fs.s3a;
+
+import org.apache.hadoop.fs.BlockLocation;
+import org.apache.hadoop.fs.LocatedFileStatus;
+
+/**
+ * {@link LocatedFileStatus} extended to also carry ETag and object version ID.
+ */
+public class S3LocatedFileStatus extends LocatedFileStatus {
+  private final String eTag;
+  private final String versionId;
+
+  public S3LocatedFileStatus(S3AFileStatus status, BlockLocation[] locations,
+      String eTag, String versionId) {
+    super(status, locations);
+    this.eTag = eTag;
+    this.versionId = versionId;
+  }
+
+  public String getETag() {
+    return eTag;
+  }
+
+  public String getVersionId() {
+    return versionId;
+  }
+
+  // equals() and hashCode() overridden to avoid FindBugs warning.
+  // Base implementation is equality on Path only, which is still appropriate.
```

Comment:
Is it appropriate, though? The whole point of this is to recognize when the file has changed in ways we haven't been looking at. Feels like comparing 2 FileStatuses should start warning if other metadata is wrong.

Reply:
It depends on perspective, but I thought it was appropriate given that LocatedFileStatus doesn't use anything else in its equality comparison, nor does the base FileStatus.

I was worried if I changed it that potentially something depending on that sort of equality would be broken when it received one of these statuses that have a different type of equality definition.

I'll do some experimentation to see if I can find any places this equals() is actually being hit to see how it is used.

Reply:
So far I've been unable to identify any place this equals() would actually be leveraged, at least within hadoop-aws itself.  I'm still hesitant to change it so let me know what you think.